storage/pin/pin.go

```diff
@@ -0,0 +1,400 @@
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package pin
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"sync"
+
+	"github.com/ethersphere/swarm/api"
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/state"
+	"github.com/ethersphere/swarm/storage"
+	"github.com/ethersphere/swarm/storage/localstore"
+)
+
+const (
+	Version            = "1.0"
+	WorkerChanSize     = 8             // Max no of goroutines when walking the file tree
+	SwarmPinHeaderName = "x-swarm-pin" // Presence of this in header indicates pinning required
+)
+
+// FileInfo is the struct that stores the information about pinned files
+// A map of this is stored in the state DB
+type FileInfo struct {
+	isRaw      bool
+	fileSize   uint64
+	pinCounter uint64
+}
+
+// API is the main object which implements all things pinning.
+type API struct {
+	db          *localstore.DB
+	api         *api.API
+	fileParams  *storage.FileStoreParams
+	tag         *chunk.Tags
+	hashSize    int
+	state       state.Store         // the state store used to store info about pinned files
+	pinnedFiles map[string]FileInfo // stores the root hashes and other info. about the pinned files
+}
+
+// NewAPI creates a API object that is required for pinning and unpinning
+func NewApi(lstore *localstore.DB, stateStore state.Store, params *storage.FileStoreParams, tags *chunk.Tags, api *api.API) *API {
+	hashFunc := storage.MakeHashFunc(storage.DefaultHash)
+	pinnedFiles := make(map[string]FileInfo)
+	err := loadPinnedFilesInfo(pinnedFiles, stateStore)
+	if err != nil {
+		log.Error("Error loading pinned files from state store.", "err", err)
+		return nil
+	}
+	return &API{
+		db:          lstore,
+		api:         api,
+		fileParams:  params,
+		tag:         tags,
+		hashSize:    hashFunc().Size(),
+		state:       stateStore,
+		pinnedFiles: pinnedFiles,
+	}
+}
+
+// PinFiles is used to pin a RAW file or a collection (which hash manifest's)
+// to the local Swarm node. It takes the root hash as the argument and walks
+// down the merkle tree and pin all the chunks that are encountered on the
+// way. It pins both data chunk and tree chunks. The pre-requisite is that
+// the file should be present in the local database. This function is called
+// from two places 1) Just after the file is uploaded 2) anytime after
+// uploading the file using the pin command. This function can pin both
+// encrypted and non-encrypted files.
+func (p *API) PinFiles(rootHash string, isRaw bool, credentials string) error {
+	addr, err := hex.DecodeString(rootHash)
+	if err != nil {
+		log.Error("Error decoding root hash", "rootHash", rootHash, "err", err)
+		return err
+	}
+
+	hasChunk, err := p.db.Has(context.TODO(), chunk.Address(p.removeDecryptionKeyFromChunkHash(addr)))
```

Comment:
Some day if the Pin and Unpin HTTP api is implemented... we could use that context. For now... this is for placeholder.. May be this can be changed to Context.Background() for now.

Reply:
I agree, thanks.